quickwit-storage/src/storage_with_upload_cache.rs

```diff
@@ -0,0 +1,885 @@
+            )));
+        }
+
+        let mut disk_cache = LruCache::unbounded();
+        for (path, size) in cache_state.items {
+            disk_cache.put(path, size);
+        }
+
+        Ok(Self {
+            remote_storage,
+            local_storage,
+            local_storage_root: local_storage_root_path,
+            disk_capacity: cache_state.disk_capacity,
+            disk_cache: Mutex::new(WrappedCache {
+                lru: LruCache::unbounded(),
+                num_bytes,
+                num_files,
+            }),
+        })
+    }
+
+    /// Check if an incomming item will exceed the capacity once inserted.
+    fn exceeds_capacity(&self, num_bytes: usize, num_files: usize) -> bool {
+        self.disk_capacity.exceeds_capacity(num_bytes, num_files)
+    }
+
+    /// Persist the state of the entire cache.
+    ///
+    /// Takes a previously locked guard for easier mutext management.
+    async fn save_state<'a>(
+        &self,
+        wrapped_cache: &MutexGuard<'_, WrappedCache>,
+    ) -> StorageResult<()> {
+        let mut cache_items = vec![];
+        for (path, size) in wrapped_cache.lru.iter() {
+            cache_items.push((path.clone(), *size));
+        }
+
+        let cache_state = CacheState {
+            local_storage_uri: self.local_storage.uri(),
+            disk_capacity: self.disk_capacity,
+            items: cache_items,
+        };
+
+        let file_path = self.local_storage_root.join(CACHE_STATE_FILE_NAME);
+        let content: Vec<u8> = serde_json::to_vec(&cache_state)
+            .map_err(|err| StorageErrorKind::InternalError.with_error(err))?;
+        atomic_write(&file_path, &content)?;
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl Storage for StorageWithUploadCache {
+    async fn put(&self, path: &Path, payload: PutPayload) -> StorageResult<()> {
+        self.remote_storage.put(path, payload.clone()).await?;
+
+        let mut locked_disk_cache = self.disk_cache.lock().await;
+        let payload_length = payload.len().await? as usize;
+        if self.exceeds_capacity(payload_length, locked_disk_cache.num_files + 1) {
+            return Ok(());
+        }
+
+        if let Some(item_num_bytes) = locked_disk_cache.lru.pop(&path.to_path_buf()) {
+            if self.local_storage.delete(path).await.is_err() {
+                return Ok(());
+            }
+            locked_disk_cache.num_bytes -= item_num_bytes;
+            locked_disk_cache.num_files -= 1;
+        }
+
+        while self.exceeds_capacity(
+            locked_disk_cache.num_bytes + payload_length,
+            locked_disk_cache.num_files + 1,
+        ) {
+            if let Some((item_path, item_num_bytes)) = locked_disk_cache.lru.pop_lru() {
+                if self
+                    .local_storage
+                    .delete(item_path.as_path())
+                    .await
+                    .is_err()
+                {
+                    return Ok(());
```

Comment:
Certainly returning `Ok(())` here is very wrong. We haven't save the file yet.

Reply:
Well not sure of this. I can be more elaborate on handling the error for non-exiting files.
But not being able to proceed just means we cannot store the item in the cache and as long as we have it the remote storage, I think `Ok(())` result is fine from the `Storage` trait point of view.